kotlinx-coroutines-core/common/src/Dispatched.kt

```diff
@@ -307,7 +307,14 @@ internal fun <T> DispatchedTask<T>.resume(delegate: Continuation<T>, useMode: In
     val state = takeState()
     val exception = getExceptionalResult(state)
     if (exception != null) {
-        delegate.resumeWithExceptionMode(exception, useMode)
+        /*
+         * Recover stacktrace for non-dispatched tasks.
+         * We usually do not recover stacktrace in a `resume` as all resumes go through `DispatchedTask.run`
+         * and we recover stacktraces there, but this is not the case for a `suspend fun main()` that knows nothing about
+         * kotlinx.coroutines and DispatchedTask
+         */
+        val recovered = if (delegate is DispatchedTask<*>) exception else recoverStackTrace(exception, delegate)
```

Comment:
This `DispatchedTask<T>.resume` extension is called only in two places. Once case is when `delegate` is not a `DispatchedContinuation`, but it is also called, for example, when there is a mismatch in resume modes. There other is from `resumeUnconfined` which is similar. Note that `DispatchTask.run` always recovers stack-trace when calling resume, so it looks like this function should unconditionally recover stack-trace, too.

Replacing this line with `val recovered = recoverStackTrace(exception, delegate)` does not break stack-track recovery test.

Reply:
>Once case is when delegate is not a DispatchedContinuation, but it is also called, for example, when there is a mismatch in resume modes.

Yes, exactly. But when there is a mismatch in resume modes, `delegate` is still `DispatchedContinuation`, thus `delegate.resume` will trigger `run` and stacktrace recovery as well.

>Replacing this line with val recovered = recoverStackTrace(exception, delegate) does not break stack-track recovery test.

Sure, because of a smart enough `StackTraceRecoveryKt.mergeRecoveredTraces`. Adding unconditional recovery will only increase the amount of unnecessary work

Reply:
Indeed!

Reply:
Indeed!